File: cli/internal/ext/manifest.go
// Package ext implements Kaioken's community extension system: packages
// published as GitHub releases, installed into the per-user directory,
// tracked by a lockfile, and discovered through a community registry index.
//
// Phase 1 is deliberately code-free. A declarative extension contributes
// documents — skills in the same SKILL.md format the repo's own skills use —
// that the agent reads through its knowledge catalog. Nothing an extension
// ships is ever executed by the host; manifests declaring executable types
// (wasm, mcp) are rejected until a sandboxed runtime exists.
package ext

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// ManifestName is the file an extension repository must carry at its root.
const ManifestName = "extension.yaml"

// TypeDeclarative is the code-free capability tier: documents the agent
// reads, never code the host runs. An empty type means the same.
const TypeDeclarative = "declarative"

// TypeMCP is the subprocess tier: the extension declares an MCP server the
// host runs as a child process and whose tools the agent may call. A
// subprocess is arbitrary code, so an mcp extension installs untrusted and
// inert — nothing launches until the user explicitly trusts the exact
// installed version.
const TypeMCP = "mcp"

// TypeWasm is the sandboxed code tier: the extension ships a WASI module the
// host runs under wazero. It is sandboxed by construction — no filesystem,
// no network, no environment — unless a declared permission grants a
// capability. It still installs inert and requires per-version trust:
// reading the user's workspace is a grant worth an explicit yes.
const TypeWasm = "wasm"

// Executable reports whether an extension type runs code and therefore
// participates in the trust/tool machinery.
func Executable(t string) bool { return t == TypeMCP || t == TypeWasm }

// MCPConfig is how an mcp extension names its server process.
type MCPConfig struct {
	// Command is resolved on PATH when it is a bare name; a path containing
	// separators must stay inside the extension's install directory unless it
	// is absolute. It runs with the install directory as working directory.
	Command string   `yaml:"command"`
	Args    []string `yaml:"args,omitempty"`
	// Env entries are appended to the inherited environment.
	Env map[string]string `yaml:"env,omitempty"`
}

// WasmConfig is how a wasm extension names its module.
type WasmConfig struct {
	// Entry is the module path relative to the package root, e.g.
	// "dist/plugin.wasm". It must stay inside the package.
	Entry string `yaml:"entry"`
}

// CommandDecl is one user-invokable command a wasm extension contributes,
// surfaced in the TUI as /x <ext> <name>.
type CommandDecl struct {
	Name        string `yaml:"name"`
	Description string `yaml:"description,omitempty"`
}

// Manifest describes one extension. It lives at the root of the extension's
// repository and travels inside every release archive, so the installed copy
// always states what it is and what it needs.
type Manifest struct {
	// ID is "owner.name": two lowercase kebab-case segments, mirroring
	// VS Code's publisher.name convention so two authors can publish
	// extensions with the same short name without colliding.
	ID string `yaml:"id"`
	// Name is the human-readable title shown in listings.
	Name string `yaml:"name"`
	// Version is a strict MAJOR.MINOR.PATCH and must match the release tag
	// (a mismatch installs with a warning rather than failing).
	Version     string `yaml:"version"`
	Description string `yaml:"description,omitempty"`
	Author      string `yaml:"author,omitempty"`
	// Repo is the informational "owner/name" origin; the lockfile records
	// where the archive actually came from.
	Repo string `yaml:"repo,omitempty"`
	// Type is the capability tier: "declarative" (or empty) and "mcp";
	// "wasm" is reserved for a future sandboxed tier.
	Type string `yaml:"type,omitempty"`
	// MCP names the server process an mcp extension runs. Required for type
	// mcp, forbidden otherwise — a declarative extension must not smuggle in
	// an executable payload.
	MCP *MCPConfig `yaml:"mcp,omitempty"`
	// Wasm names the module a wasm extension runs. Required for type wasm,
	// forbidden otherwise.
	Wasm *WasmConfig `yaml:"wasm,omitempty"`
	// Permissions are the capabilities a wasm module asks for, granted (or
	// refused) as a set at trust time. Unknown permissions fail installation
	// outright: a capability the host cannot grant must never be silently
	// dropped, or the plugin breaks in confusing ways at runtime.
	Permissions []string `yaml:"permissions,omitempty"`
	// MinKaiokenVersion refuses installation on hosts older than the
	// extension needs, checked again at load through the lock entry's
	// installed manifest.
	MinKaiokenVersion string `yaml:"minKaiokenVersion,omitempty"`
	// Hooks are agent lifecycle events the extension wants forwarded into
	// the sandbox (tool_call, tool_result, agent_end, …). Wasm only, and
	// only dispatched once the installed version is trusted.
	Hooks []string `yaml:"hooks,omitempty"`
	// Commands are user-invokable entry points, run via /x. Wasm only.
	Commands []CommandDecl `yaml:"commands,omitempty"`
}

// LoadManifest reads and validates dir/extension.yaml.
func LoadManifest(dir string) (*Manifest, error) {
	raw, err := os.ReadFile(filepath.Join(dir, ManifestName))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("no %s found — this is not a Kaioken extension", ManifestName)
		}
		return nil, err
	}
	var m Manifest
	if err := yaml.Unmarshal(raw, &m); err != nil {
		return nil, fmt.Errorf("parsing %s: %w", ManifestName, err)
	}
	if err := m.Validate(); err != nil {
		return nil, err
	}
	return &m, nil
}

// Validate checks the fields an installer must be able to trust: the id
// shape (it becomes a directory name), the version (it drives updates), and
// the type (the security boundary of this whole phase).
func (m *Manifest) Validate() error {
	if err := validateID(m.ID); err != nil {
		return err
	}
	if strings.TrimSpace(m.Name) == "" {
		return fmt.Errorf("extension %s has no name", m.ID)
	}
	if _, err := parseSemver(m.Version); err != nil {
		return fmt.Errorf("extension %s: %w", m.ID, err)
	}
	switch m.Type {
	case "", TypeDeclarative:
		if m.MCP != nil {
			return fmt.Errorf("extension %s: declarative extensions must not declare an mcp server", m.ID)
		}
		if m.Wasm != nil {
			return fmt.Errorf("extension %s: declarative extensions must not declare a wasm module", m.ID)
		}
	case TypeMCP:
		if m.MCP == nil || strings.TrimSpace(m.MCP.Command) == "" {
			return fmt.Errorf("extension %s: mcp extensions must declare mcp.command", m.ID)
		}
		if m.Wasm != nil {
			return fmt.Errorf("extension %s: mcp extensions must not declare a wasm module", m.ID)
		}
	case TypeWasm:
		if m.Wasm == nil || strings.TrimSpace(m.Wasm.Entry) == "" {
			return fmt.Errorf("extension %s: wasm extensions must declare wasm.entry", m.ID)
		}
		entry := m.Wasm.Entry
		if !strings.HasSuffix(entry, ".wasm") || safeRel(entry) != nil || filepath.IsAbs(entry) {
			return fmt.Errorf("extension %s: wasm.entry must be a relative .wasm path inside the package", m.ID)
		}
		if m.MCP != nil {
			return fmt.Errorf("extension %s: wasm extensions must not declare an mcp server", m.ID)
		}
	default:
		return fmt.Errorf("extension type %q is not supported yet — only declarative, mcp and wasm extensions install in this version", m.Type)
	}
	if len(m.Permissions) > 0 && m.Type != TypeWasm {
		return fmt.Errorf("extension %s: permissions apply to wasm extensions only", m.ID)
	}
	for _, p := range m.Permissions {
		if !knownPermission(p) {
			return fmt.Errorf("extension %s: permission %q is not supported yet", m.ID, p)
		}
	}
	if strings.TrimSpace(m.MinKaiokenVersion) != "" {
		if _, err := parseSemver(m.MinKaiokenVersion); err != nil {
			return fmt.Errorf("extension %s: invalid minKaiokenVersion: %w", m.ID, err)
		}
	}
	if len(m.Hooks) > 0 && m.Type != TypeWasm {
		return fmt.Errorf("extension %s: hooks require a wasm extension in this version", m.ID)
	}
	for _, h := range m.Hooks {
		if !validHookName(h) {
			return fmt.Errorf("extension %s: unknown hook %q", m.ID, h)
		}
	}
	if len(m.Commands) > 0 && m.Type != TypeWasm {
		return fmt.Errorf("extension %s: commands require a wasm extension in this version", m.ID)
	}
	for _, c := range m.Commands {
		if !kebab(c.Name) {
			return fmt.Errorf("extension %s: command name %q must be lowercase kebab-case", m.ID, c.Name)
		}
	}
	return nil
}

// validateID enforces the "owner.name" shape. The id doubles as the install
// directory name, so the character set is deliberately narrow.
func validateID(id string) error {
	segs := strings.Split(id, ".")
	if len(segs) != 2 || !kebab(segs[0]) || !kebab(segs[1]) {
		return fmt.Errorf("invalid extension id %q: want owner.name in lowercase kebab-case", id)
	}
	return nil
}

// kebab reports whether s is non-empty lowercase kebab-case: [a-z0-9-] with
// no leading or trailing dash.
func kebab(s string) bool {
	if s == "" || s[0] == '-' || s[len(s)-1] == '-' {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if (c < 'a' || c > 'z') && (c < '0' || c > '9') && c != '-' {
			return false
		}
	}
	return true
}

File: .kaioken_v2/packages/ext/src/manifest.ts
import { readFile } from "node:fs/promises";
import { isAbsolute, join } from "node:path";
import { parse as parseYaml } from "yaml";
import { parseSemver } from "./semver.js";

/**
 * What an extension says it is.
 *
 * The manifest is a security boundary before it is a description: the id
 * becomes a directory name, the version drives updates, and the type decides
 * whether anything the package ships is ever executed. So validation is strict
 * and stated in one place — a manifest that passes here is one the installer is
 * allowed to trust about those three things, and nothing else.
 */

export const MANIFEST_NAME = "extension.yaml";

/** Documents only. Nothing an extension of this type ships is ever run. */
export const TYPE_DECLARATIVE = "declarative";
/** A server process the host launches and whose tools the agent may call. */
export const TYPE_MCP = "mcp";
/** A WebAssembly module the host runs in the engine's own sandbox. */
export const TYPE_WASM = "wasm";

export type ExtensionType = "declarative" | "mcp" | "wasm";

export interface McpConfig {
	/** Resolved on PATH when it is a bare name. Runs in the install directory. */
	command: string;
	args?: string[];
	env?: Record<string, string>;
}

export interface WasmConfig {
	/** Module path relative to the package root, e.g. "dist/plugin.wasm". */
	entry: string;
}

export interface CommandDecl {
	name: string;
	description?: string;
}

export interface Manifest {
	/** "owner.name": two lowercase kebab-case segments. */
	id: string;
	name: string;
	version: string;
	description?: string;
	author?: string;
	/** Informational "owner/repo" origin; the lockfile records the real source. */
	repo?: string;
	type: ExtensionType;
	mcp?: McpConfig;
	wasm?: WasmConfig;
	/** Capabilities a wasm module asks for, granted as a set at trust time. */
	permissions?: string[];
	minKaiokenVersion?: string;
	/** User-invokable entry points, run through `/x`. Wasm and mcp only. */
	commands?: CommandDecl[];
}

/**
 * Capabilities the host knows how to grant.
 *
 * An unknown permission fails installation outright rather than being dropped:
 * a capability that silently disappears leaves a plugin failing at runtime for
 * a reason nobody can see from either side.
 */
const KNOWN_PERMISSIONS = new Set(["read_repo", "network"]);

const ID_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*\.[a-z0-9]+(?:-[a-z0-9]+)*$/;

export async function loadManifest(dir: string): Promise<Manifest> {
	let raw: string;
	try {
		raw = await readFile(join(dir, MANIFEST_NAME), "utf8");
	} catch {
		throw new Error(`no ${MANIFEST_NAME} found — this is not a Kaioken extension`);
	}

	let parsed: unknown;
	try {
		parsed = parseYaml(raw);
	} catch (error) {
		throw new Error(`parsing ${MANIFEST_NAME}: ${(error as Error).message.split("\n")[0]}`);
	}
	return validateManifest(parsed);
}

export function validateManifest(value: unknown): Manifest {
	if (!value || typeof value !== "object" || Array.isArray(value)) {
		throw new Error(`${MANIFEST_NAME} is not a mapping`);
	}
	const record = value as Record<string, unknown>;

	const id = text(record.id);
	// The id becomes a path segment on disk, so it may not be one.
	if (!ID_PATTERN.test(id)) {
		throw new Error(`extension id "${id}" must be "owner.name" in lowercase kebab-case`);
	}
	const name = text(record.name);
	if (!name) throw new Error(`extension ${id} has no name`);

	const version = text(record.version);
	if (!parseSemver(version)) {
		throw new Error(`extension ${id}: "${version}" is not a MAJOR.MINOR.PATCH version`);
	}

	const declared = text(record.type) || TYPE_DECLARATIVE;
	if (declared !== TYPE_DECLARATIVE && declared !== TYPE_MCP && declared !== TYPE_WASM) {
		throw new Error(`extension type "${declared}" is not supported — declarative, mcp or wasm`);
	}
	const type = declared as ExtensionType;

	const mcp = record.mcp ? readMcp(id, record.mcp) : undefined;
	const wasm = record.wasm ? readWasm(id, record.wasm) : undefined;

	// The cross-checks are what stop a declarative extension — the tier a user
	// installs without thinking about it — from smuggling in a payload.
	if (type === TYPE_DECLARATIVE) {
		if (mcp) throw new Error(`extension ${id}: declarative extensions must not declare an mcp server`);
		if (wasm) throw new Error(`extension ${id}: declarative extensions must not declare a wasm module`);
	}
	if (type === TYPE_MCP) {
		if (!mcp) throw new Error(`extension ${id}: mcp extensions must declare mcp.command`);
		if (wasm) throw new Error(`extension ${id}: mcp extensions must not declare a wasm module`);
	}
	if (type === TYPE_WASM) {
		if (!wasm) throw new Error(`extension ${id}: wasm extensions must declare wasm.entry`);
		if (mcp) throw new Error(`extension ${id}: wasm extensions must not declare an mcp server`);
	}

	const permissions = strings(record.permissions);
	if (permissions.length > 0 && type !== TYPE_WASM) {
		throw new Error(`extension ${id}: permissions apply to wasm extensions only`);
	}
	for (const permission of permissions) {
		if (!KNOWN_PERMISSIONS.has(permission)) {
			throw new Error(`extension ${id}: permission "${permission}" is not supported`);
		}
	}

	const minimum = text(record.minKaiokenVersion);
	if (minimum && !parseSemver(minimum)) {
		throw new Error(`extension ${id}: invalid minKaiokenVersion "${minimum}"`);
	}

	const commands = readCommands(record.commands);
	if (commands.length > 0 && type === TYPE_DECLARATIVE) {
		throw new Error(`extension ${id}: declarative extensions contribute documents, not commands`);
	}

	return {
		id,
		name,
		version,
		type,
		...(text(record.description) ? { description: text(record.description) } : {}),
		...(text(record.author) ? { author: text(record.author) } : {}),
		...(text(record.repo) ? { repo: text(record.repo) } : {}),
		...(mcp ? { mcp } : {}),
		...(wasm ? { wasm } : {}),
		...(permissions.length > 0 ? { permissions } : {}),
		...(minimum ? { minKaiokenVersion: minimum } : {}),
		...(commands.length > 0 ? { commands } : {}),
	};
}

/** Does this type run code, and therefore need per-version trust? */
export function isExecutable(type: ExtensionType): boolean {
	return type === TYPE_MCP || type === TYPE_WASM;
}

function readMcp(id: string, value: unknown): McpConfig {
	if (!value || typeof value !== "object") throw new Error(`extension ${id}: mcp must be a mapping`);
	const record = value as Record<string, unknown>;
	const command = text(record.command);
	if (!command) throw new Error(`extension ${id}: mcp extensions must declare mcp.command`);
	const env: Record<string, string> = {};
	if (record.env && typeof record.env === "object") {
		for (const [key, entry] of Object.entries(record.env as Record<string, unknown>)) {
			if (typeof entry === "string") env[key] = entry;
		}
	}
	return {
		command,
		...(strings(record.args).length > 0 ? { args: strings(record.args) } : {}),
		...(Object.keys(env).length > 0 ? { env } : {}),
	};
}

function readWasm(id: string, value: unknown): WasmConfig {
	if (!value || typeof value !== "object") throw new Error(`extension ${id}: wasm must be a mapping`);
	const entry = text((value as Record<string, unknown>).entry);
	// The entry is joined onto the install directory, so a path that escapes it
	// would let a package name any file on the machine as its module.
	if (!entry.endsWith(".wasm") || isAbsolute(entry) || entry.split(/[\\/]/).includes("..")) {
		throw new Error(`extension ${id}: wasm.entry must be a relative .wasm path inside the package`);
	}
	return { entry };
}

function readCommands(value: unknown): CommandDecl[] {
	if (!Array.isArray(value)) return [];
	const out: CommandDecl[] = [];
	for (const entry of value) {
		if (!entry || typeof entry !== "object") continue;
		const record = entry as Record<string, unknown>;
		const name = text(record.name);
		if (!name) continue;
		out.push({ name, ...(text(record.description) ? { description: text(record.description) } : {}) });
	}
	return out;
}

function text(value: unknown): string {
	return typeof value === "string" ? value.trim() : "";
}

function strings(value: unknown): string[] {
	return Array.isArray(value) ? value.filter((entry): entry is string => typeof entry === "string") : [];
}
